src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -49,15 +50,15 @@ use crate::task::{ActorId, CreateMviewProgress};
 /// - [`ArrangementBackfillExecutor`] can reside on a different CN, so it can be scaled
 ///   independently.
 /// - To synchronize upstream shared buffer, it is initialized with a [`ReplicatedStateTable`].
-pub struct ArrangementBackfillExecutor<S: StateStore> {
+pub struct ArrangementBackfillExecutor<S: StateStore, SD: ValueRowSerde> {
     /// Upstream table
-    upstream_table: ReplicatedStateTable<S>,
+    upstream_table: ReplicatedStateTable<S, SD>,
 
     /// Upstream with the same schema with the upstream table.
     upstream: BoxedExecutor,
 
     /// Internal state table for persisting state of backfill state.
-    state_table: ReplicatedStateTable<S>,
+    state_table: StateTable<S>,
```

Comment:
> However, if there is no replication, how does it read the non-ckpt data from this state table in arrangement backfill?

This is intentional. Arrangement backfill in this PR is basically `no_shuffle_backfill` but with vnode level snapshot read.

In future PR I will add replication and amend the dispatcher shuffling logic for it. Not adding in this PR, because it will make it harder to review.

May take a look at `#future-work` in PR description for next steps.